modules/text2vec-bigram/main.go

```diff
@@ -0,0 +1,187 @@
+//
+
+package t2vbigram
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/weaviate/weaviate/entities/models"
+	"github.com/weaviate/weaviate/entities/modulecapabilities"
+	"github.com/weaviate/weaviate/entities/moduletools"
+	"github.com/weaviate/weaviate/entities/schema"
+	"github.com/weaviate/weaviate/usecases/modulecomponents/additional"
+	"github.com/weaviate/weaviate/usecases/modulecomponents/arguments/nearText"
+	libvectorizer "github.com/weaviate/weaviate/usecases/vectorizer"
+)
+
+const Name = "text2vec-bigram"
+
+func New() *BigramModule {
+	return &BigramModule{}
+}
+
+type BigramModule struct {
+	vectors                      map[string][]float32
+	GraphqlProvider              modulecapabilities.GraphQLArguments
+	Searcher                     modulecapabilities.Searcher
+	NearTextTransformer          modulecapabilities.TextTransform
+	logger                       logrus.FieldLogger
+	AdditionalPropertiesProvider modulecapabilities.AdditionalProperties
+}
+
+func (m *BigramModule) Arguments() map[string]modulecapabilities.GraphQLArgument {
+	return map[string]modulecapabilities.GraphQLArgument{}
+}
+
+func (m *BigramModule) Name() string {
+	return Name
+}
+
+func (m *BigramModule) Type() modulecapabilities.ModuleType {
+	return modulecapabilities.Text2Vec
+}
+
+func (m *BigramModule) Init(ctx context.Context, params moduletools.ModuleInitParams) error {
+	m.logger = params.GetLogger()
+	return nil
+}
+
+func (m *BigramModule) InitExtension(modules []modulecapabilities.Module) error {
+	return nil
+}
+
+func (m *BigramModule) InitVectorizer(ctx context.Context, timeout time.Duration, logger logrus.FieldLogger) error {
+	return nil
+}
+
+func (m *BigramModule) VectorizableProperties(cfg moduletools.ClassConfig) (bool, []string, error) {
+	return true, []string{}, nil
+}
+
+func (m *BigramModule) InitAdditionalPropertiesProvider() error {
+	return nil
+}
+
+func (m *BigramModule) RootHandler() http.Handler {
+	return nil
+}
+
+func (m *BigramModule) VectorizeObject(ctx context.Context, obj *models.Object, cfg moduletools.ClassConfig) ([]float32, models.AdditionalProperties, error) {
+	var text string
+	for _, prop := range obj.Properties.(map[string]interface{}) {
+		text += prop.(string)
+	}
+	vector, error := m.VectorizeInput(ctx, text, cfg)
+	return vector, nil, error
+}
+
+func (m *BigramModule) MetaInfo() (map[string]interface{}, error) {
+	return nil, nil
+}
+
+func (m *BigramModule) AdditionalProperties() map[string]modulecapabilities.AdditionalProperty {
+	return additional.NewText2VecProvider().AdditionalProperties()
+}
+
+func text2vector(input string) ([]float32, error) {
+	bigramcount := map[string]int{}
+	for i := 0; i < len(input)-1; i++ {
+		bigram := input[i : i+2]
+		bigramcount[bigram]++
+	}
+
+	vector := make([]float32, 256*256)
+	for bigram, count := range bigramcount {
+		index := int(bigram[0]) * int(bigram[1])
+		vector[index] = float32(count)
+	}
```

Comment:
that is a very big vector - could we make it smaller? Maybe by doing `vector[index%256]` or something like this?

In general I like the idea and would use it for tests in the python client. But a length of 65500 is a bit much

Reply:
Do we have vector compression?  I thought we did, but maybe I'm confusing that with PQ

Reply:
> Do we have vector compression? I thought we did, but maybe I'm confusing that with PQ. Most of the vector will be zeros, and will compress very nicely.

Only PQ, I think in general it does not bring that much

Reply:
The usecase is fun.  Also, if not for using, why have 64k vectors?  And what's the point of having a large hard drive if you don't fill it up with data?  Free disk space is wasted money!

More seriously, we don't have an embedded vectoriser, which makes testing painful.  Wouldn't it be nice to have a single binary weaviate that just works for everything, without downloading and installing python and all the modules?  It doesn't really matter if 1million vectors blows out to 65Gb, we have terabyte drives.

I looked at our code, and it looks like we don't compress our vectors.  That seems like a good feature to have, if we support 64k vectors.  And depending on the speeds of disk and cpu, it might not actually be slower.  A vector that is mostly zeros will compress _really_ well.

This is also a form of sparse vectors, which are a popular subject right now (splade).

Also, as you noted, we can reduce the length of the vectors by throwing away some data, but we could improve on bucketing (using %256) by including a dictionary of common bigrams and limiting vectors to the top, e.g. 2048 pairs.

Plenty of opportunities for interesting experiments.

Reply:
>  It doesn't really matter if 1million vectors blows out to 65Gb, we have terabyte drives.

Can you tell my boss to buy me a new laptop please?

On a serious note:
In the clients (and probably in weaviate itself, too) we often need a vectorizer that does something, but transfering 65k vectors in tests takes too long. If we could downsize the vectors and still have somewhat meaningful results it would really be great :)

Having said that, I don't want to stand in the way if you think the current setup is what you need, it just doesn't rub _my_ itch in its current form :)

Reply:
I'm sorry, I didn't mean to exclude you.  I think we could do what you want by, e.g. mapping all input to 26 letters, and then using bigrams.  Would 26^2 be an acceptable vector length?

I'll add a few more options this weekend (it's my weekend project).  Then you can see which one fits your needs.

Reply:
> I'm sorry, I didn't mean to exclude you.

Didn't understood it that way, you're scratching your itch :) 

>  Would 26^2 be an acceptable vector length?

yep that sounds good! Maybe it could be as simple as having a `vector_length` parameter that simply buckets the entries (eg doing %vector_length)